### DISCUSSION TITLE
Issue loading Bioconductor packages on Emu

### DISCUSSION THREAD

**Discussion by aspencoyle** [RobertsLab/resources#1186] [Q&A] [answered]
I'm working on Emu right now, and am having some issues trying to install Bioconductor packages (specifically WGCNA) when trying to run my Hematodinium analysis script. The problem seems to be twofold - first, RStudio seems to have problems connecting to the internet, and second, there seems to be a permissions error.

Running R version 3.5.1 (which seems to be the latest for Linux - checked for updates, said no updates available).

Anyway, here's some code snippets and their output:

Started off by running `library(BiocManager)` a few times. The first two returned `Warning message:
Bioconductor version cannot be validated; no internet
  connection? `. However, the third loaded the library just fine.

I then tried to run `BiocManager::install("WGCNA")`, which returned `Error: Bioconductor version cannot be validated; no internet
  connection?`, regardless of how many times I ran it.
  
  Tried to load another package to test if that was the problem with `library(tidyverse)`, and that ran with no issues.
  
  Did some googling, and one suggestion was to run 
```
config <- "https://bioconductor.org/config.yaml"
readLines(config)
```
to see if that connected to bioconductor.org. It did not - it returned 

```
Error in file(con, "r") : 
  cannot open the connection to 'https://bioconductor.org/config.yaml'
In addition: Warning message:
In file(con, "r") :
  URL 'https://bioconductor.org/config.yaml': status was 'Problem with the SSL CA cert (path? access rights?)'
```

This seems to indicate that Emu is having a problem connecting to Bioconductor, but I can't find anything on how to solve it! I tried installing BiocManager from scratch with `install.packages("BiocManager")`, which returned:

```
 % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  259k  100  259k    0     0   690k      0 --:--:-- --:--:-- --:--:--  690k
* installing *source* package ‘BiocManager’ ...
** package ‘BiocManager’ successfully unpacked and MD5 sums checked
** R
** inst
** byte-compile and prepare package for lazy loading
** help
*** installing help indices
** building package indices
** installing vignettes
** testing if installed package can be loaded
* DONE (BiocManager)

The downloaded source packages are in
	‘/tmp/Rtmp2GQAEd/downloaded_packages’
Updating HTML index of packages in '.Library'
Warning in install.packages :
  cannot create file '/opt/microsoft/ropen/3.5.1/lib64/R/doc/html/packages.html', reason 'Permission denied'
Warning in install.packages :
  cannot update HTML package index
```

This really isn't urgent - I'm just going to switch over from Emu to another computer in the lab and see if that runs. But thought I'd bring it up in case it is a reoccuring issue or has an easy solution!

**Comment by kubu4** [comment]
I've gone ahead and installed WGCNA. Give it a shot and let me know if you can run it.

Not sure what the deal was with the weird internet message.

The permissions issue is likely due related to whichever user group your Emu profile is assigned to. I'd likely guess that your user account doesn't have permission to install packages. I'll try to figure this out tomorrow, but if you'll anticipate needing to install more packages, I'd recommend trying to do so in the srlab account. If you need the PW, hit me up in Slack.

**Reply by aspencoyle** [reply]
Fantastic, thanks a bunch - the package is now loading just fine! Would you be able to do the same for the DESeq2 package as well? That should be the last package I need!

And funny thing is, I am using the srlab account! No idea what's going on there...

**Reply by kubu4** [reply]
Post command to install Deseq2 and I should be able to get it installed. Sorry about this. Is a tad surprising, but I'm not sure how many people (anyone?) Have tried installing R packages in that computer.

**Reply by aspencoyle** [reply]
Yeah sure! It should just be `BiocManager::install("DESeq2")`

Thanks again for getting that fixed!

**Reply by kubu4** [reply]
This has not gone smoothly and will require much more work to get DESeq2 installed (weird compiler error). I'll try to tackle it tomorrow. Sorry!

**Comment by kubu4** [comment, marked as answer]
Alrighty, I've modified things a bit (installed Microsoft Open R v4.0.2), which means you'll need to re-run some installations, like BiocManager and WGSNA yourself (sorry!), but things should operate more smoothly.

The biggest change is that when a user goes to install a package for the first time, the system now prompts the user to use a personal library (@afcoyle I'd recommend you log in to your dedicated user account instead of the srlab account, btw):

![image](https://user-images.githubusercontent.com/4514104/115562492-62ef9180-a26b-11eb-8f59-bf21860b8ce6.png)


If it's not obvious, the user should answer `yes` to that question. Now, all libraries should install without an write permission junk from now on!

So, give it a rip (in your user account!) and let me know how things go.

To install DESeq2, you might need to run the following (from the command line):

1. `R` (this starts R)
2. `Sys.setenv("PKG_CXXFLAGS"="-std=c++11")` This tells R to use a different compiler than the default.
3.  `BiocManager::install("DESeq2")` Installs DESeq2

NOTE: This has currently only been implemented on Emu. I will get Roadrunner updated in a bit.
